workshop/exercise-6/README.md

```diff
@@ -1 +1,117 @@
+
+```yaml
+apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
+metadata:
+  name: virtual-service-guestbook
+spec:
+  hosts:
+  - guestbook
+  http:
+  - route:
+    - destination:
+        host: guestbook
+        subset: v1
+```
+
+```yaml
+apiVersion: networking.istio.io/v1alpha3
+kind: DestinationRule
+metadata:
+  name: destination-rule-guestbook
+spec:
+  host: guestbook
+  subsets:
+  - name: v1
+    labels:
+      version: v1
+  - name: v2
+    labels:
+      version: v2
+```
+
+The `VirtualService` defines a rule that captures all HTTP traffic going to hostname `guestbook` and routes that traffic pods of the guestbook service with label "version: v1".  A subset or version of a route destination is identified with a reference to a named service subset which must be declared in a corresponding `DestinationRule`.
+
+To begin incrementally routing traffic to the newer version of the guestbook service, the original `VirtualService` rule can be modified to following:
+
+```yaml
+apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
+metadata:
+  name: virtual-service-guestbook
+spec:
+  hosts:
+  - guestbook
+  http:
+    - route:
+      - destination:
+          host: guestbook
+          subset: v1
+        weight: 80
+      - destination:
+          host: guestbook
+          subset: v2
+        weight: 20
+```
+
+In the modified rule, the routed traffic is split between two different subsets of the guestbook service.  In this manner, traffic to the modernized version 2 of guestbook is controlled on a percentage basis to limit the impact of any undiscovered bugs.  This rule can be modified over time until eventually all traffic is directed to the newer version of the service.
+
+### A/B Testing with Istio
+A/B testing is a method of performing identical tests against two separate service versions in order to determine which performs better.  The original `VirtualService` rule above can be modified to the following:
+
+```yaml
+apiVersion: networking.istio.io/v1alpha3
+kind: VirtualService
+metadata:
+  name: virtual-service-guestbook
+spec:
+  hosts:
+  - guestbook
+  http:
+  - match:
+    - headers:
+        cookie:
+          regex: "^(.*?;)?(user=jason)(;.*)?$"
+    route:
+    - destination:
+        host: guestbook
+        subset: v2
+  - route:
+    - destination:
+        host: guestbook
+        subset: v1
+```
+
+In Istio `VirtualService` rules, there can be only one rule for each service and therefore when defining multiple `http` blocks, the order in which they are defined in the yaml matters.  With the modified rule, incoming requests going to hostname `guestbook` and contain a Cookie header with `user=jason` will go to the newer version of guestbook.  All other requests fallthrough to the next block, which routes all traffic to the original version of guestbook.  This way the same performance tests and functionality tests can be used to run against version 1 and version 2 just by appending a header.
+
```

Comment:
this may require code change to the app to be able to do a A/B testing.  Perhaps a quick A/B testing can be browser based as the app doesn't use user right now.

Reply:
If this were a demo, and we just wanted the people taking the course to be able to visually see difference between version 1 and version 2 of the service in action by viewing the guestbook in different browsers, this would be the better way to go.  But since there is not a demo portion for this section, A/B testing best practices would probably not follow this approach because in theory you would want to test both versions of the services on both browsers to verify they behave the same whether they are viewed in Firefox/Safari/etc.  I could set it to a more arbitrary header and value like:

```
  - match:
    - headers:
        istio-test-header:
          exact: "version2"
```